// downloader-m3u8/build/esm/src/downloadM3u8/downloadM3u8TsFileTask.js
import { DownLoad } from 'downloader-util';
import { Task } from '../task/taskUtil';
export default class DownloadM3u8TsFileTask extends Task {
    fileUrl;
    filePath;
    headers;
    retry = 100;
    downloadTask;
    constructor(fileUrl, filePath, headers) {
        super();
        this.fileUrl = fileUrl;
        this.filePath = filePath;
        this.headers = headers;
    }
    async task() {
        await this.doTask();
    }
    getFileUrl() {
        return this.fileUrl;
    }
    getFilepath() {
        return this.filePath;
    }
    async doTask() {
        return new Promise((res, rej) => {
            this.downloadTask = new DownLoad({
                url: this.fileUrl,
                filePath: this.filePath,
                headers: this.headers,
                timeout: 10000,
                onFailed: (error) => {
                    if (this.retry > 0 && error != '用户手动停止，停止下载') {
                        this.retry--;
                        res(this.doTask());
                    }
                    else {
                        rej(error);
                    }
                },
                onSuccess: () => {
                    res('');
                },
            });
            this.downloadTask.start();
        });
    }
    stop() {
        if (this.downloadTask) {
            this.downloadTask.stop();
        }
    }
}
