// downloader-m3u8/build/cjs/test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("./index");
function downloadFile(_a) {
    var url = _a.url, name = _a.name;
    return __awaiter(this, void 0, void 0, function () {
        var e_1;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, index_1.downloadM3u8FileToMp4({
                            m3u8Url: url,
                            filePath: "../tmp/",
                            title: name,
                            onProgress: function (progress) {
                                console.log("\u8FDB\u5EA6---" + progress * 100 + "%");
                            },
                        })];
                case 1:
                    _b.sent();
                    return [3 /*break*/, 3];
                case 2:
                    e_1 = _b.sent();
                    console.log('下载失败！');
                    console.log(e_1);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    });
}
function downloadFile1(_a) {
    var url = _a.url, name = _a.name;
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_b) {
            return [2 /*return*/, new Promise(function (res, rej) {
                    var task = new index_1.DownloadM3u8FileMp4({
                        m3u8Url: url,
                        filePath: "../tmp/",
                        title: name,
                        headers: {
                            referer: 'https://missav.com/',
                            host: 'vz-8bd03a41-60e.b-cdn.net',
                            accept: '*/*',
                            'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36',
                        },
                    });
                    task.onProgress(function (progress) {
                        console.log(name + "\u8FDB\u5EA6---" + progress * 100 + "%");
                    });
                    task.onFinish(function () {
                        console.log(name + "\u5DF2\u5B8C\u6210");
                        res('');
                    });
                    task.onFail(function (e) {
                        console.log(name + "--" + e);
                        rej(e);
                    });
                    task.start();
                    // setTimeout(() => {
                    //     console.log('10s之后停止!')
                    //     task.stop();
                    // }, 10000)
                })];
        });
    });
}
function main() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, downloadFile({
                        url: 'https://vs02.520call.me/files/mp4/t/tPYUY.m3u8?t=1633880449',
                        // url: 'https://v.xboku.com/20200813/TZFNhKYK/index.m3u8',
                        name: 'CLUB-639 只對我這個陰暗圖書委員溫柔 偷拍短髮美乳女子。上課時也持續內射交尾[有碼中文字幕]',
                    })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
main();
