// downloader-util/build/cjs/test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
process.env['NODE_TLS_REJECT_UNAUTHORIZED'] = '0';
var index_1 = require("./src/index");
var path = require('path');
var mypath = path.resolve(__dirname, 'tmp/a.m3u8');
console.log(mypath);
var url = 'https://vz-8bd03a41-60e.b-cdn.net/b8010124-bc2c-4e46-96c7-a3880c8f4026/1280x720/video.m3u8';
var task = new index_1.DownLoad({
    url: url,
    filePath: mypath,
    threadCount: 20,
    type: 1,
    headers: {
        referer: 'https://missav.com/',
        host: 'vz-8bd03a41-60e.b-cdn.net',
        accept: '*/*',
        'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36',
    },
    onFailed: function (error) {
        console.log("\u5931\u8D25\u539F\u56E0\uFF01\uFF01\uFF01:" + error);
    },
    onSuccess: function () {
        console.log('文件下载成功！');
    },
    onProgress: function (progress) {
        console.log("\u8FDB\u5EA6:" + Math.floor(progress * 100) + "%");
    },
});
task.start();
// const fs = require('fs-extra');
// const fd = fs.openSync('./tmp/a.mp4', 'a');
// var buffer = new Buffer('加这么一段文字！！！！！');
// console.log(buffer.length);
// console.log(buffer)
// fs.writeSync(fd, buffer, 0, buffer.length, 100001);
