// downloader-m3u8/build/esm/src/task/taskUtil.js
/**
 * 一个简易的任务池，
 */
var TASK_STATUS;
(function (TASK_STATUS) {
    TASK_STATUS[TASK_STATUS["PENDING"] = 0] = "PENDING";
    TASK_STATUS[TASK_STATUS["DONGING"] = 1] = "DONGING";
    TASK_STATUS[TASK_STATUS["DONE"] = 2] = "DONE";
})(TASK_STATUS || (TASK_STATUS = {}));
export class Task {
    taskStatus = TASK_STATUS.PENDING;
    connection = null;
    taskIsSuccess = false;
    excute() {
        this.donging();
        const taskHandle = this.task();
        if (taskHandle instanceof Promise) {
            taskHandle
                .then(() => {
                this.taskIsSuccess = true;
            })
                .catch((error) => {
                this.taskIsSuccess = false;
                console.log(error);
            })
                .finally(() => {
                this.done();
            });
        }
        else {
            this.done();
        }
    }
    setConnection(connection) {
        this.connection = connection;
    }
    donging() {
        this.taskStatus = TASK_STATUS.DONGING;
    }
    done() {
        this.taskStatus = TASK_STATUS.DONE;
        this.connection && this.connection.disconnect(this);
    }
    getStatus() {
        return this.taskStatus;
    }
    stop() {
    }
}
export class TaskPool {
    taskMaxNum;
    hasFinishCount;
    poolState;
    constructor(taskMaxNum, hasFinishCount) {
        this.taskMaxNum = taskMaxNum;
        this.hasFinishCount = hasFinishCount;
        this.poolState = 0;
    }
    taskList = [];
    doningList = [];
    doneList = [];
    finishListeners = [];
    progressListeners = [];
    stopListeners = [];
    addTask(task) {
        if (this.poolState != 0) {
            console.log('当前任务池已经停止运行');
            return;
        }
        if (task.getStatus() === TASK_STATUS.PENDING) {
            this.taskList.push(task);
        }
        if (this.taskList.length === 1 &&
            this.doningList.length < this.taskMaxNum) {
            // 当有一个任务时且有空余的task执行
            this.excute();
        }
    }
    removeTask(task) {
        if (task.getStatus() !== TASK_STATUS.DONGING) {
            task.done(); // 不从task栈中去除，只改变状态，改变状态后，轮到此任务会直接跳过
        }
    }
    excute() {
        if (this.poolState !== 0) {
            return;
        }
        if (this.doningList.length < this.taskMaxNum) {
            const task = this.taskList.shift();
            if (task && task.getStatus() !== TASK_STATUS.DONE) {
                task.setConnection(this);
                this.doningList.push(task);
                task.excute();
            }
            else {
                // 任务完成
                if (this.doningList.length === 0) {
                    this.finishListeners.forEach((lis) => {
                        if (lis) {
                            lis(this.doneList);
                        }
                    });
                    this.doneList = [];
                    this.finishListeners = [];
                    this.progressListeners = [];
                }
            }
        }
    }
    disconnect(task) {
        if (this.poolState !== 0) {
            return;
        }
        // 将任务从doningList中删除
        const doningList = this.doningList;
        doningList.find((taskItem, index) => {
            if (taskItem === task) {
                doningList.splice(index, 1);
                return true;
            }
        });
        this.doneList.push(task);
        this.doProgress(task);
        this.excute();
    }
    doProgress(task) {
        if (this.poolState !== 0) {
            return;
        }
        const doneLen = this.doneList.length + this.hasFinishCount;
        const allLen = doneLen + this.doningList.length + this.taskList.length || 1;
        const progress = doneLen / allLen;
        this.progressListeners.forEach((listener) => {
            if (listener) {
                listener(progress, task);
            }
        });
    }
    onProgress(listener) {
        this.progressListeners.push(listener);
    }
    addFinishListener(listener) {
        this.finishListeners.push(listener);
    }
    addStopListener(listener) {
        this.stopListeners.push(listener);
    }
    isEmpty() {
        return this.taskList.length == 0 && this.doningList.length == 0;
    }
    clear() {
        this.taskList = [];
        this.doningList.forEach((task) => {
            task.stop();
        });
        this.doneList = [];
        this.doningList = [];
    }
    stop() {
        this.poolState = 1;
        this.clear();
        this.stopListeners.forEach((listener) => {
            if (listener) {
                listener('任务已停止');
            }
        });
    }
}
